### ISSUE TITLE
RFC 6242 Chunked Framing Compliant?

### ISSUE THREAD

**Issue by ctdeitrich** [ncclient/ncclient#155]
Is ncclient v0.5.2 compliant with the Chunked Framing Mechanism described in RFC6242 section 4.1?  I have written a NETCONF server in Python, and it works with netopeer for v1.1, however when I connect using ncclient it appears that the v1.1 message is not chunked.  It appears to me, however, that RFC6242 section 4.1 is mandatory?

Thanks,
Toby

**Comment by ctdeitrich** [comment]
Here is an excerpt from RFC6242:

4.1.  Framing Protocol

   The previous version of this document defined the character sequence
   "]]>]]>" as a message separator, under the assumption that it could
   not be found in well-formed XML documents.  However, this assumption
   is not correct.  It can legally appear in XML attributes, comments,
   and processing instructions.  In order to solve this problem, and at
   the same time be compatible with existing implementations, this
   document defines the following framing protocol.

   The <hello> message MUST be followed by the character sequence
   ]]>]]>.  Upon reception of the <hello> message, the receiving peer's
   SSH Transport layer conceptually passes the <hello> message to the
   Messages layer.  **If the :base:1.1 capability is advertised by both
   peers, the chunked framing mechanism (see Section 4.2) is used for
   the remainder of the NETCONF session.**  Otherwise, the old end-of-
   message-based mechanism (see Section 4.3) is used.

**Comment by mirceaulinic** [comment]
@ctdeitrich I am not a maintainer of this library (assuming there's still someone looking here once in a while...), but my understanding is that ncclient is not compliant with that mechanism. Looking into https://github.com/ncclient/ncclient/blob/master/ncclient/transport/ssh.py#L108:

```python
        if MSG_DELIM in buf.read().decode('UTF-8'):
            buf.seek(0)
            msg, _, remaining = buf.read().decode('UTF-8').partition(MSG_DELIM)
            msg = msg.strip()
            if sys.version < '3':
                self._dispatch_message(msg.encode())
            else:
                self._dispatch_message(msg)
            # create new buffer which contains remaining of old buffer
            self._buffer = StringIO()
            self._buffer.write(remaining.encode())
            self._parsing_pos10 = 0
        else:
            # handle case that MSG_DELIM is split over two chunks
            self._parsing_pos10 = buf.tell() - MSG_DELIM_LEN
            if self._parsing_pos10 < 0:
                self._parsing_pos10 = 0
```

Seems that it stops reading at the first occurrence of ```]]>]]>```.
However, this is not a problem as the network devices escape all special characters and if the XML would need to contain ```]]>]]>``` would be transformed into ```]]&gt;]]&gt;``` so there are no risks.

E.g.:

Raw reply of an arbitrary request:

```
mircea@re1.edge01.sjc01> show system boot-messages
Copyright (c) 1996-2016, Juniper Networks, Inc.
All rights reserved.
Copyright (c) 1992-2007 The FreeBSD Project.
Copyright (c) 1979, 1980, 1983, 1986, 1988, 1989, 1991, 1992, 1993, 1994
       	The Regents of the University of California. All rights reserved.
FreeBSD is a registered trademark of The FreeBSD Foundation.
JUNOS 14.2R6.5 #0: 2016-04-05 03:35:25 UTC
    builder@orth.juniper.net:/volume/build/junos/14.2/release/14.2R6.5/obj/amd64/junos/bsd/kernels/JUNIPER64/kernel
Timecounter "i8254" frequency 1193182 Hz quality 0
CPU: Intel(R) Xeon(R) CPU           C5518  @ 1.73GHz (1729.11-MHz K8-class CPU)
  Origin = "GenuineIntel"  Id = 0x106e4  Stepping = 4
  Features=0xbfebfbff<FPU,VME,DE,PSE,TSC,MSR,PAE,MCE,CX8,APIC,SEP,MTRR,PGE,MCA,CMOV,PAT,PSE36,CLFLUSH,DTS,ACPI,MMX,FXSR,SSE,SSE2,SS,HTT,TM,PBE>
  Features2=0x9ce3bd<SSE3,RSVD2,MON,DS_CPL,VMX,EST,TM2,<b9>,CX16,<b14>,<b15>,<b18>,<b19>,<b20>,<b23>>
  AMD Features=0x28100800<SYSCALL,NX,RDTSCP,LM>
  AMD Features2=0x1<LAHF>
  Cores per package: 8
  Logical CPUs per core: 2
real memory  = 17179869184 (16384 MB)
avail memory = 16603320320 (15834 MB)
(...snip...)
```

Which is transmitted via netconf after escaping ```>``` and others:

```xml
<?xml version="1.0" encoding="UTF-8"?><nc:rpc xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0"><nc:command format="xml">show system boot-messages</nc:command></nc:rpc>
<rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:junos="http://xml.juniper.net/junos/14.2R6/junos" xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0">
<output>
Copyright (c) 1996-2016, Juniper Networks, Inc.
All rights reserved.
Copyright (c) 1992-2007 The FreeBSD Project.
Copyright (c) 1979, 1980, 1983, 1986, 1988, 1989, 1991, 1992, 1993, 1994
       	The Regents of the University of California. All rights reserved.
FreeBSD is a registered trademark of The FreeBSD Foundation.
JUNOS 14.2R6.5 #0: 2016-04-05 03:35:25 UTC
    builder@orth.juniper.net:/volume/build/junos/14.2/release/14.2R6.5/obj/amd64/junos/bsd/kernels/JUNIPER64/kernel
Timecounter "i8254" frequency 1193182 Hz quality 0
CPU: Intel(R) Xeon(R) CPU           C5518  @ 1.73GHz (1729.11-MHz K8-class CPU)
  Origin = "GenuineIntel"  Id = 0x106e4  Stepping = 4
  Features=0xbfebfbff&lt;FPU,VME,DE,PSE,TSC,MSR,PAE,MCE,CX8,APIC,SEP,MTRR,PGE,MCA,CMOV,PAT,PSE36,CLFLUSH,DTS,ACPI,MMX,FXSR,SSE,SSE2,SS,HTT,TM,PBE&gt;
  Features2=0x9ce3bd&lt;SSE3,RSVD2,MON,DS_CPL,VMX,EST,TM2,&lt;b9&gt;,CX16,&lt;b14&gt;,&lt;b15&gt;,&lt;b18&gt;,&lt;b19&gt;,&lt;b20&gt;,&lt;b23&gt;&gt;
  AMD Features=0x28100800&lt;SYSCALL,NX,RDTSCP,LM&gt;
  AMD Features2=0x1&lt;LAHF&gt;
  Cores per package: 8
  Logical CPUs per core: 2
real memory  = 17179869184 (16384 MB)
avail memory = 16603320320 (15834 MB)
(...snip...)
</output>
</rpc-reply>
```

Not that the line ```AMD Features2=0x1<LAHF>``` is transmitted as ```AMD Features2=0x1&lt;LAHF&gt;``` etc.

In that case my hunch is that you should enhance your server implementation to escape before returning the XML reply.

**Comment by leopoul** [comment]
Hi,
Is your server opensource? I'd be very much interested to try and debug. Current version supports v1.1 chunked framing and has been tested against netopeer and Cisco ASR devices.
So, handshake begins with a hello msg 
```
# send a HELLO msg using v1.0 EOM markers.
validated_element(data, tags='{urn:ietf:params:xml:ns:netconf:base:1.0}hello')
data = "%s%s"%(data, MSG_DELIM)
```
(https://github.com/ncclient/ncclient/blob/master/ncclient/transport/ssh.py#L523) 
using the "]]>]]>" delimiter 
```
MSG_DELIM = "]]>]]>"
```
(https://github.com/ncclient/ncclient/blob/master/ncclient/transport/ssh.py#L37). 

If "urn:ietf:params:netconf:base:1.1" is found in server capabilities, then for the remainder of the session the chunk framing mechanism is used 
```
if self._server_capabilities:
                            if 'urn:ietf:params:netconf:base:1.1' in self._server_capabilities and 'urn:ietf:params:netconf:base:1.1' in self._client_capabilities:
                                logger.debug("Selecting netconf:base:1.1 for encoding")
self._parse11()
```
(https://github.com/ncclient/ncclient/blob/master/ncclient/transport/ssh.py#L510)

For the definition of _parse11, check here: https://github.com/ncclient/ncclient/blob/master/ncclient/transport/ssh.py#L136. 

Could you set loglevel to debug and paste the output? I'd like to make sure that the chunking mechanism on the client side works as expected and that 1.1 is indeed the parsing mode selected by the client.

Thanks in advance,

L.

**Comment by ctdeitrich** [comment]
Hi @leopoul,

Thanks much for your reply.  Unfortunately, our server isn't Open Source.  However, we looked into the issue and found that ncclient is sensitive to trailing whitespace in the hello message content.  We removed the whitespace from our hello messages and now our server is working with ncclient.  Not sure what the NETCONF standard requires regarding trailing whitespace, but might be worth doing a strip() in the client code.

Thanks!
Toby
